File: DSA_python/DSA04024 LŨY THỪA MA TRẬN 2.py
def submit():
    n, k = map(int, input().split())
    A = []
    for i in range(n):
        A.append(list(map(int, input().split())))

    def multiply(A, B, n):
        C = [[0]*n for _ in range(n)]
        for i in range(n):
            for j in range(n):
                for k in range(n):
                    C[i][j] = (C[i][j] + A[i][k]*B[k][j])%MOD

        return C

    def power(A, n, times):
        ans = [[0]*n for _ in range(n)]
        for i in range(n):
            for j in range(n):
                if i == j:
                    ans[i][j] = 1
        
        while times:
            if times&1:
                ans = multiply(ans, A, n)
            A = multiply(A, A, n)
            times >>= 1
        return ans
    
    A = power(A, n, k)
    total = 0

    for i in range(n):
        total = (total + A[i][n-1])%MOD
    
    print(total%MOD)


MOD = 10**9+7

test = int(input())
for _ in range(test):
    submit()

File: DSA/DSA04024 LŨY THỪA MA TRẬN 2.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1e9 + 7;

void multiply(vector<vector<long long>> &A, vector<vector<long long>> &B, int n)
{
    vector<vector<long long>> C(n, vector<long long> (n));

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            for (int k = 0; k < n; k++)
            {
                C[i][j] += (A[i][k]*B[k][j])%MOD;
            }
        }
    }

    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            A[i][j] = C[i][j]%MOD;
        }
    }
}

vector<vector<long long>> power(vector<vector<long long>> A, int n, int times)
{
    vector<vector<long long>> ans(n, vector<long long> (n));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            if (i == j)
            {
                ans[i][j] = 1;
            }
            else
            {
                ans[i][j] = 0;
            }
        }
    }
    
    while (times)
    {
        if (times & 1)
        {
            multiply(ans, A, n);
        }
        multiply(A, A, n);
        times >>= 1;
    }
    
    return ans;
}

void submit()
{
    int n, k;
    cin >> n >> k;

    vector<vector<long long>> A(n, vector<long long> (n));
    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < n; j++)
        {
            cin >> A[i][j];
        }
    }

    A = power(A, n, k);
    long long total = 0;

    for (int i = 0; i < n; i++)
    {
        total += A[i][n-1]%MOD;
    }
    
    cout << total%MOD;

    cout << endl;
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    int test;
    cin >> test;
    while (test--)
    {
        submit();
    }
    

    return 0;
}